### PR TITLE
Clang-CUDA and HIP Support

### PR DESCRIPTION
**Pull request by fthaler**
Changes to enable Clang-CUDA and AMD HIP compilation:
- Disabled building of cpp_bindgen tests.
- Added `GT_CUDA_COMPILATION_TYPE` option to choose between NVCC-CUDA, Clang-CUDA and HIPCC-AMDGPU compilation options.
- Introduced hip_wrappers.hpp, trivially mapping CUDA names to HIP names using macros.
- Minor code changes to make all code Clang-CUDA- and HIP-compatible.

Performance-relevant changes:
- Unroll factor for vertical advection tuned on NVIDIA V100, AMD Mi50 (slight performance regression on older GPUs).
- More compiler-friendly implementation of storage_info constructor.
- Simple horizontal diffusion benchmark is now using parallel execution strategy.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -45,7 +45,14 @@ FetchContent_Declare(
...
 # TODO remove in GT 2.0
 set(CPP_BINDGEN_GT_LEGACY ON)
+if (GT_USE_CLANG_CUDA)
+    set(build_testing_ ${BUILD_TESTING})
+    set(BUILD_TESTING OFF)
+endif()
 FetchContent_MakeAvailable(cpp_bindgen)
+if (GT_USE_CLANG_CUDA)
+    set(BUILD_TESTING ${build_testing_})
+endif()
```
Comment (fthaler) [inline CMakeLists.txt]: Cpp_bindgen’s CMake setup does not yet support Clang-CUDA, sot its unit tests are disabled (the tests inside GridTools using cpp_bindgen are still enabled, as they work perfectly).

File: docs_src/manuals/getting_started/code/CMakeLists.txt [hunk 2]
```diff
@@ -2,6 +2,8 @@ cmake_minimum_required(VERSION 3.14.5)
 
 project(GridTools-laplacian LANGUAGES CXX)
 
+set(CMAKE_CXX_EXTENSIONS OFF)
```
Comment (fthaler) [inline docs_src/manuals/getting_started/code/CMakeLists.txt]: Required due to Clang not supporting -std=gnu++14 on AMD GPUs.

File: include/gridtools/meta/is_meta_class.hpp [hunk 3]
```diff
@@ -1,28 +0,0 @@
...
-
-namespace gridtools {
-    namespace meta {
-        /**
-         *   meta class concept check
-         */
-        template <class, class = void>
-        struct is_meta_class : std::false_type {};
-        template <class T>
-        struct is_meta_class<T, void_t<curry<T::template apply>>> : std::true_type {};
```
Comment (fthaler) [inline include/gridtools/meta/is_meta_class.hpp]: Clang trunk has compilation issues here, but the function isn’t used anyway, so I removed it.

File: unit_tests/common/test_cuda_is_ptr.cu [hunk 4]
```diff
@@ -31,18 +31,6 @@ TEST(test_is_gpu_ptr, cuda_ptr_is_cuda_ptr) {
     EXPECT_EQ(cudaSuccess, cudaGetLastError());
 }
 
-TEST(test_is_gpu_ptr, cudaMallocHost_ptr_is_not_cuda_ptr) {
-    auto testee = std::shared_ptr<double>(
-        []() {
-            double *res;
-            GT_CUDA_CHECK(cudaMallocHost(&res, sizeof(double)));
```
Comment (fthaler) [inline unit_tests/common/test_cuda_is_ptr.cu]: cuda/hipMallocHost is not supported by HIP, but test seems irrelevant for us anyway, so I removed it.

File: CMakeLists.txt [hunk 5]
```diff
@@ -45,7 +45,14 @@ FetchContent_Declare(
 # Enable legacy mode for CPP_BINDGEN for users of GT 1.0
 # TODO remove in GT 2.0
 set(CPP_BINDGEN_GT_LEGACY ON)
+if (GT_USE_CLANG_CUDA)
```
Comment (havogt) [inline CMakeLists.txt]: Please disable unconditionally.

**Comment by fthaler** [comment]
launch strgrid

**Comment by fthaler** [comment]
launch icgrid

**Comment by fthaler** [comment]
launch perftest

File: cmake/flags.cmake [hunk 6]
```diff
@@ -15,6 +15,14 @@ option(GT_GCL_ONLY "If on only library is build but not the examples and tests"
 
 option(GT_TESTS_ICOSAHEDRAL_GRID "compile tests for icosahedral grids" OFF )
 
+CMAKE_DEPENDENT_OPTION(
+    GT_USE_CLANG_CUDA "Compile CUDA code with Clang"
+    OFF "GT_ENABLE_BACKEND_CUDA" OFF)
+
+CMAKE_DEPENDENT_OPTION(
+    GT_USE_HIP "Use AMD HIP instead of CUDA"
```
Comment (havogt) [inline cmake/flags.cmake]: See also first comment. This looks like a weird chain of dependencies.

File: cmake/definitions.cmake [hunk 7]
```diff
@@ -41,21 +41,42 @@ endif()
 target_compile_definitions(gridtools INTERFACE BOOST_PP_VARIADICS=1)
 if(CUDA_AVAILABLE)
     target_compile_definitions(gridtools INTERFACE GT_USE_GPU)
-    if(${CMAKE_CUDA_COMPILER_VERSION} VERSION_LESS 9.0)
-        message(FATAL_ERROR "CUDA 8.X or lower is not supported")
+    if(GT_USE_HIP)
```
[the remark this answers is not in the document]
Comment (fthaler) [reply cmake/definitions.cmake]: Alternative implementation now available: uses GT_CUDA_COMPILATION_TYPE. What do you think?

File: include/gridtools/stencil_composition/backend_cuda/fused_mss_loop_cuda.hpp [hunk 8]
```diff
@@ -55,7 +55,10 @@ namespace gridtools {
                     k_caches_type<Mss> k_caches;
                     auto mixed_ptr = hymap::device::merge(k_caches.ptr(), wstd::move(ptr));
                     tuple_util::device::for_each(
-                        [&](int_t size, auto info) {
+                        [&](int_t size, auto info) GT_FORCE_INLINE_LAMBDA {
+#ifdef __clang__
+#pragma unroll 3
```
Comment (havogt) [inline include/gridtools/stencil_composition/backend_cuda/fused_mss_loop_cuda.hpp]: add comment how this value was chosen, to give a pointer to the next developer when it would make sense to change it.

File: unit_tests/storage/storage_cuda/test_data_view.cu [hunk 9]
```diff
@@ -24,7 +24,11 @@ template <typename View>
 __global__ void mul2(View s) {
     bool correct_dims = (s.template total_length<0>() == c_x) && (s.template total_length<1>() == c_y) &&
                         (s.template total_length<2>() == c_z);
+#ifdef __HIPCC__
+    bool correct_size = (s.padded_total_length() == 16 * c_y * c_z);
```
Comment (havogt) [inline unit_tests/storage/storage_cuda/test_data_view.cu]: Can we instead query the alignment from  the storage_info, would make the test more readable.

File: unit_tests/storage/storage_cuda/test_data_view.cu [hunk 10]
```diff
@@ -68,7 +72,11 @@ TEST(DataViewTest, Simple) {
 
     ASSERT_TRUE((si.padded_total_length() == dv.padded_total_length()));
 
+#ifdef __HIPCC__
+    ASSERT_TRUE(si.index(1, 0, 1) == c_y * 16 + 1);
```
Comment (havogt) [inline unit_tests/storage/storage_cuda/test_data_view.cu]: same

File: unit_tests/storage/test_storage_facility.cpp [hunk 11]
```diff
@@ -56,20 +56,25 @@ struct static_type_tests {
 template <>
 struct static_type_tests<backend::cuda> {
     using storage_traits_t = storage_traits<backend::cuda>;
+#ifdef __HIPCC__
+    using alignment_t = alignment<16>;
```
Comment (havogt) [inline unit_tests/storage/test_storage_facility.cpp]: I don't like it.

File: unit_tests/storage/test_storage_facility.cpp [hunk 12]
```diff
@@ -56,20 +56,25 @@ struct static_type_tests {
...
 
     /*########## STORAGE INFO CHECKS ########## */
     // storage info check
     typedef storage_traits_t::storage_info_t<0, 3, halo<1, 2, 3>> storage_info_ty;
     GT_STATIC_ASSERT(
         (is_storage_info<storage_info_ty>::type::value), "is_storage_info metafunction is not working anymore");
-    GT_STATIC_ASSERT((std::is_same<storage_info_ty,
-                         storage_info<0, layout_map<2, 1, 0>, halo<1, 2, 3>, alignment<32>>>::type::value),
+    GT_STATIC_ASSERT(
+        (std::is_same<storage_info_ty, storage_info<0, layout_map<2, 1, 0>, halo<1, 2, 3>, alignment_t>>::type::value),
```
Comment (havogt) [inline unit_tests/storage/test_storage_facility.cpp]: I don't think this test is very useful anyway, but maybe we just shouldn't check the alignment (or query the alignment from the storage info as proposed above).

File: unit_tests/storage/test_storage_facility.cpp (hunk 11 above)
Comment (fthaler) [reply unit_tests/storage/test_storage_facility.cpp]: So what’s your proposal here?

File: unit_tests/storage/test_storage_facility.cpp (hunk 12 above)
Comment (fthaler) [reply unit_tests/storage/test_storage_facility.cpp]: But it looks likes it’s the exact purpose of this test to check those types? So either it should include alignment or otherwise we can drop the test, right?

Comment (havogt) [reply unit_tests/storage/test_storage_facility.cpp]: I don't think the test makes much sense.

Comment (fthaler) [reply unit_tests/storage/test_storage_facility.cpp]: I removed it…

**Comment by fthaler** [comment]
launch strgrid

**Comment by fthaler** [comment]
launch icgrid

**Comment by fthaler** [comment]
launch perftest
